Manually document System.IO.BinaryWriter, StreamWriter, StringWriter, TextWriter

carlossanlop opened the pull request:
Changes for TextWriter, StreamWriter, BinaryWriter and StringWriter.

Owners are @JeremyKuhne and I.
@mairaw and @rpetrusha for language review.

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -1414,9 +1414,19 @@ This member is an explicit interface member implementation. It can be used only
+
+## Remarks
+
+The text representation of the specified value is produced by calling the <xref:System.ReadOnlySpan&lt;System.Char&gt;.ToString%2A?displayProperty=nameWithType> method.
```
```suggestion
The text representation of the specified value is produced by calling the [ReadOnlySpan<Char>.ToString](xref:System.ReadOnlySpan%601.ToString%2A) method.
```

carlossanlop replied:
@rpetrusha oh, I see what you are doing here. I'll edit the malformed one in the first file to look like this one.

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -3086,9 +3106,21 @@ This member is an explicit interface member implementation. It can be used only
+
+## Remarks
+
+The text representation of the specified value is produced by calling the <xref:System.ReadOnlySpan&lt;System.Char&gt;.ToString%2A?displayProperty=nameWithType> method.
```
```suggestion
The text representation of the specified value is produced by calling the [ReadOnlySpan<Char>.ToString](xref:System.ReadOnlySpan%601.ToString%2A) method.
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -1570,9 +1580,17 @@ This member is an explicit interface member implementation. It can be used only
+
+## Remarks
+
+Equivalent to `Write(stringBuilder.ToString())`, but uses the `StringBuilder.GetChunks()` method to avoid creating the intermediate string.
```
```suggestion
This method is equivalent to `Write(stringBuilder.ToString())`, but it uses the <xref:System.Text.StringBuilder.GetChunks%2A?displayProperty=nameWithType>  method to avoid creating the intermediate string.
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -2389,10 +2407,10 @@ This member is an explicit interface member implementation. It can be used only
-        <param name="cancellationToken">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="buffer">The char region of memory to write to the text stream.</param>
```
```suggestion
        <param name="buffer">The character region of memory to write to the text stream.</param>
```

carlossanlop replied:
memory region:
```suggestion
        <param name="buffer">The character memory region to write to the text stream.</param>
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -2389,10 +2407,10 @@ This member is an explicit interface member implementation. It can be used only
-        <returns>To be added.</returns>
+        <param name="buffer">The char region of memory to write to the text stream.</param>
+        <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="P:System.Threading.CancellationToken.None" />.</param>
+        <summary>Asynchronously writes a char region of memory to the text stream.</summary>
```
```suggestion
        <summary>Asynchronously writes a character region of memory to the text stream.</summary>
```

carlossanlop replied:
```suggestion
        <summary>Asynchronously writes a character memory region to the text stream.</summary>
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -3086,9 +3106,21 @@ This member is an explicit interface member implementation. It can be used only
-        <summary>To be added.</summary>
-        <remarks>To be added.</remarks>
+        <param name="buffer">The char span value to write to the text stream.</param>
+        <summary>Writes the text representation of a char span to the text stream, followed by a line terminator.</summary>
```
```suggestion
        <summary>Writes the text representation of a character span to the text stream, followed by a line terminator.</summary>
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -4144,11 +4191,19 @@ This member is an explicit interface member implementation. It can be used only
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="buffer">The char memory region to write to the text stream.</param>
```
```suggestion
        <param name="buffer">The character memory region to write to the text stream.</param>
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -4144,11 +4191,19 @@ This member is an explicit interface member implementation. It can be used only
-        <remarks>To be added.</remarks>
+        <param name="buffer">The char memory region to write to the text stream.</param>
+        <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="P:System.Threading.CancellationToken.None" />.</param>
+        <summary>Asynchronously writes the text representation of a char memory region to the text stream, followed by a line terminator.</summary>
```
```suggestion
        <summary>Asynchronously writes the text representation of a character memory region to the text stream, followed by a line terminator.</summary>
```

rpetrusha commented on xml/System.IO/TextWriter.xml:
```diff
@@ -4231,17 +4294,18 @@ This member is an explicit interface member implementation. It can be used only
+
+The line terminator is defined by the <xref:System.IO.TextWriter.CoreNewLine> field.
+
+The <xref:System.IO.TextWriter> class is an abstract class. Therefore, you do not instantiate it in your code. For an example of using the <xref:System.IO.TextWriter.WriteLineAsync%2A> method, see the <xref:System.IO.StreamWriter.WriteLineAsync%2A?displayProperty=nameWithType> method.
```
This seems like an unusual place for this comment.  I think that it can be deleted, or possibly it could be removed to the remarks of the WriteAsync \<MemberGroup>, if one exists.

carlossanlop replied:
I see what you mean. Yes, I'll move it to the MemberGroup for WriteLineAsync, in line 3930.
```suggestion
```

carlossanlop commented on xml/System.IO/StreamWriter.xml:
```diff
@@ -1552,10 +1552,18 @@ For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common
+
+## Remarks
+
+See <xref:System.IO.TextWriter.Write(System.String,System.Object)> for a description of the composite formatting capabilities offered.
```
@rpetrusha is it ok to reference its parent's implementation instead of copy-pasting the whole remark?

rpetrusha replied:
Yes, it is, @carlossanlop.

carlossanlop commented on xml/System.IO/StreamWriter.xml:
```diff
@@ -2021,9 +2053,21 @@ For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common
+
+## Remarks
+
+The text representation of the specified value is produced by calling the <xref:System.ReadOnlySpan&lt;System.Char&gt;.ToString%2A?displayProperty=nameWithType> method.
```
Hey @rpetrusha I think this suggestion is malformed. The `[` is not closed. Can you please double check it?

carlossanlop commented on xml/System.IO/StreamWriter.xml:
```diff
@@ -1917,10 +1949,10 @@ For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common
-        <param name="cancellationToken">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="buffer">The char region of memory to write to the stream.</param>
```
I changed region of memory -> memory region, to match the rest of the suggestions below.

```suggestion
        <param name="buffer">The char memory region to write to the stream.</param>
```

carlossanlop commented on xml/System.IO/StreamWriter.xml:
```diff
@@ -1917,10 +1949,10 @@ For a list of common I/O tasks, see [Common I/O Tasks](~/docs/standard/io/common
-        <returns>To be added.</returns>
+        <param name="buffer">The char region of memory to write to the stream.</param>
+        <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="P:System.Threading.CancellationToken.None" />.</param>
+        <summary>Asynchronously writes a char region of memory to the stream.</summary>
```
Same:

```suggestion
        <summary>Asynchronously writes a character memory region to the stream.</summary>
```

carlossanlop commented on xml/System.IO/StringWriter.xml:
```diff
@@ -1273,10 +1273,10 @@
-        <param name="cancellationToken">To be added.</param>
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
+        <param name="buffer">The char region of memory to write to the string.</param>
```
Region of memory -> Memory region:

```suggestion
        <param name="buffer">The character memory region to write to the string.</param>
```

carlossanlop commented on xml/System.IO/TextWriter.xml:
```diff
@@ -4231,17 +4294,18 @@ This member is an explicit interface member implementation. It can be used only
+The line terminator is defined by the <xref:System.IO.TextWriter.CoreNewLine> field.
+
+The <xref:System.IO.TextWriter> class is an abstract class. Therefore, you do not instantiate it in your code. For an example of using the <xref:System.IO.TextWriter.WriteLineAsync%2A> method, see the <xref:System.IO.StreamWriter.WriteLineAsync%2A?displayProperty=nameWithType> method.
+
```
```suggestion
```

carlossanlop commented on xml/System.IO/TextWriter.xml:
```diff
@@ -3881,7 +3927,7 @@ This member is an explicit interface member implementation. It can be used only
       </AssemblyInfo>
       <Docs>
-        <summary>Writes data followed by a line terminator asynchronously to the text string or stream.</summary>
+        <summary>Asynchronously writes data to the text stream, followed by a line terminator.</summary>
```
```suggestion
        <summary>Asynchronously writes data to the text stream, followed by a line terminator.</summary>
        <remarks>
          <format type="text/markdown">
            <![CDATA[

## Remarks

The <xref:System.IO.TextWriter> class is an abstract class. Therefore, you do not instantiate it in your code. For an example of using the <xref:System.IO.TextWriter.WriteLineAsync%2A> method, see the <xref:System.IO.StreamWriter.WriteLineAsync%2A?displayProperty=nameWithType> method.

            ]]>
          </format>
        </remarks>
```

carlossanlop commented:
@rpetrusha I addressed all your suggestions. I also left a question for you. If there are no more suggestions, we could get this merged today if the build finishes successfully.

rpetrusha commented:
This looks good, @carlossanlop. I'll merge now.
